sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsPartitions.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.connector.catalog;
+
+import java.util.Map;
+
+import org.apache.spark.annotation.Experimental;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.analysis.NoSuchPartitionException;
+import org.apache.spark.sql.catalyst.analysis.PartitionAlreadyExistsException;
+
+/**
+ * A partition interface of {@link Table}.
+ * A partition is composed of identifier and properties,
+ * and properties contains metadata information of the partition.
+ * <p>
+ * These APIs are used to modify table partition identifier or partition metadata,
+ * in some cases, they will change actual value of table data as well.
+ *
+ * @since 3.0.0
+ */
+@Experimental
+public interface SupportsPartitions extends Table {
+
+    /**
+     * Create a partition in table.
+     *
+     * @param ident a new partition identifier
+     * @param properties the metadata of a partition
+     * @throws PartitionAlreadyExistsException If a partition already exists for the identifier
+     */
+    void createPartition(
+            InternalRow ident,
+            Map<String, String> properties) throws PartitionAlreadyExistsException;
+
+    /**
+     * Drop a partition from table.
+     *
+     * @param ident a partition identifier
+     * @return true if a partition was deleted, false if no partition exists for the identifier
+     */
+    Boolean dropPartition(InternalRow ident);
+
+    /**
+     * Rename a Partition from old identifier to new identifier with no metadata changed.
+     *
+     * @param oldIdent the partition identifier of the existing partition
+     * @param newIdent the new partition identifier of the partition
+     * @throws NoSuchPartitionException If the partition identifier to rename doesn't exist
+     * @throws PartitionAlreadyExistsException If the new partition identifier already exists
+     */
+    void renamePartition(
+            InternalRow oldIdent,
+            InternalRow newIdent) throws NoSuchPartitionException, PartitionAlreadyExistsException;
```

Comment:
The use case for this method still doesn't make sense to me. Why would you want to have such a limited update operation? I've never seen the ability to change values of a partition like this.

For Hive tables, this breaks some readers that depend on the file paths to have the correct partition column values, or would require moving files between directories, which is not atomic.

Reply:
Hm, you are right about this API. I can remove this API from current version if it is controversial.
BTW, how would you solve the problem that users want rename partition and remove data? Just unsupported?

Reply:
I think this is because there are different ways of thinking about partitions. If the partition you're working with is like the [example with `PARTITION europe VALUES ('France', 'Italy')`](https://www.enterprisedb.com/edb-docs/d/edb-postgres-advanced-server/user-guides/database-compatibility-for-oracle-developers-guide/9.6/Database_Compatibility_for_Oracle_Developers_Guide_v9.6.1.111.html) then it makes a lot more sense to have a rename operation.

But I think this makes less sense here because we are passing the values that identify a partition. In that example, the equivalent would be `country='France'`. In that case, a rename makes much less sense because the identifier isn't really a name in the same sense.

I think that the partitions in that example are more similar to the partitioning transforms that we configure at a table level. You might express that as a partition expression with a case statement, like this:

```sql
CREATE TABLE sales (...)
PARTITIONED BY (CASE WHEN country IN ('France', 'Italy') THEN 'Europe' WHEN ... ELSE 'Unknown' END)
```

For now, I think it makes sense to leave this out, since the identifier is not really a name. Does that sound reasonable?

Reply:
Hm, that's sounds reasonable, thanks.